### DISCUSSION TITLE
Running code before executing lambda

### DISCUSSION THREAD

**Discussion by igorg-il** [jobrunr/jobrunr#404] [Q&A]
Hello
Currently we run 
BackgroundJob.enqueue(() -> bean.doWork())

We need to add some code execution (as part of the job) before bean.doWork() is called. This code can't be added as part of the bean.

Tried the following code but it throws JobClassNotFoundException exception 
```
BackgroundJob.enqueue(() -> {
   doSomething(someArgument);   
   bean.doWork()
});
```

Tried also 
```
job.enquque.<BeanType>enquue(bean -> {
  doSomething(someArgument);   
  bean.doWork()
});
```

And got error "Jobrune only supports single enqueueing/scheduling of one method"
Is there some way to do this?
thank you

**Comment by rdehuyss** [comment]
Yes, using a JobFilter. 

Der https://www.jobrunr.io/en/documentation/#jobfilter.

If you're using I would appreciate a GitHub star (also for all the support).

**Reply by igorg-il** [reply]
First of all thank you for fast answer.
I saw filter can be defined during configuration (JobRunr.configure().withJobFilters(...)) or as annotation on job method
During configuration we don't have enough data to run pre-process code yet.
We have few jobs but we call enqueue in one place. All of the jobs require the same pre-execution code. So using annotation will be code duplication (which we are trying to prevent).
Is there any way to run filter in code just before enqueue?
thank you
P.S. added star.

**Comment by rdehuyss** [comment]
With JobFilter, you have access to the job so you can also run logic there. In the Pro version, you can dependency injection in those JobFilter's and I personally never had a need for more.

If that doesn't work, the only thing I can think of then is to compose your jobs.

E.g. wrap your job and call the wrapping method as job.

**Comment by igorg-il** [comment]
Got a big progress. But have a small issue.
When I add filters on job annotation it works (jobFilters = {MyClientFilter.class, MyJobServerFilter.class}.
But we don't want to add the same annotations for all out jobs (trying to prevent code duplication).

I added the following code which adds filters to JobRunr.configure() but filters are not called.  
What am I missing?

Thank you

```
 @Bean
    public JobRunrConfigurationResult jobRunrConfigurationResult(MongoDBStorageProvider mongoDBStorageProvider) {
        return JobRunr.configure()
                .useStorageProvider(mongoDBStorageProvider)
                .withJobFilter(new MyJobClientFilter(), new MyJobServerFilter())
                .initialize();
    }
```
In application.yaml:
```
org.jobrunr.background-job-server.enabled: true
org.jobrunr.dashboard:
    enabled: true
    port: 8000


```

**Reply by rdehuyss** [reply]
You are mixing the fluent API and the Spring style. You created a `@Bean` of type `JobRunrConfigurationResult` but that bean will not be used. If you want JobFilters, you need to use the Fluent API in the OSS version of JobRunr.
